File: client/src/lib/csvExport.ts
/**
 * Shared CSV export helper.
 *
 * Every export in the app used to join rows with `r.join(",")`, which has two
 * faults.
 *
 * It is not valid RFC 4180. Any value containing a comma, a double quote or a
 * newline silently shifts every following column, so one text field with a comma
 * in it corrupts the whole row and the reader has no way to tell.
 *
 * And nothing lined up. A heading like "Barometric Pressure (hPa)" is 25
 * characters and its values are 4, so no value sat under the name it belonged to
 * and the file could only be read in a spreadsheet, never as text. That is the
 * complaint this module exists to fix.
 *
 * Alignment rules, both deliberate:
 *
 *  - Padding goes on the RIGHT, so a value still begins immediately after its
 *    comma. Spaces inside a field are part of the field under RFC 4180, and a
 *    numeric column with LEADING spaces is what makes Excel import the column as
 *    text. Trailing spaces are trimmed when a cell is coerced to a number, so the
 *    file stays machine-readable as well as legible.
 *  - The last column is never padded. Trailing whitespace at end of line buys no
 *    alignment and some diff and lint tools flag it.
 *
 * A UTF-8 BOM is prepended because Excel assumes the system code page otherwise
 * and mangles degree signs and anything else non-ASCII.
 *
 * Property of METRON (PTY) LTD | Inteltronics
 * Developed by L.J. Esterhuizen, Inteltronics
 */

/** Quote a single field to RFC 4180. */
export function csvField(v: unknown): string {
  const s = v === null || v === undefined ? "" : String(v);
  return /[",\n\r]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

/**
 * Build an aligned, RFC 4180 quoted CSV document.
 *
 * `extraRows` are appended after a blank line and are NOT padded to the data
 * grid: provenance lines are prose, not table rows, and stretching them to the
 * column widths only misaligns them.
 */
export function buildAlignedCsv(
  headers: readonly unknown[],
  rows: readonly (readonly unknown[])[],
  extraRows: readonly string[] = [],
): string {
  const headerCells = headers.map(csvField);
  const dataCells = rows.map(r => r.map(csvField));

  const colCount = headerCells.length;
  const widths = new Array<number>(colCount).fill(0);
  for (const row of [headerCells, ...dataCells]) {
    for (let i = 0; i < colCount; i++) {
      const len = (row[i] ?? "").length;
      if (len > widths[i]) widths[i] = len;
    }
  }

  const formatRow = (row: string[]): string =>
    row
      .map((cell, i) => {
        const s = cell ?? "";
        return i === colCount - 1 ? s : s + " ".repeat(Math.max(0, widths[i] - s.length));
      })
      .join(",");

  const lines = [formatRow(headerCells), ...dataCells.map(formatRow)];
  if (extraRows.length > 0) {
    lines.push("");
    lines.push(...extraRows);
  }
  // Leading BOM so Excel reads it as UTF-8.
  return "\uFEFF" + lines.join("\n");
}

/** Trigger a browser download of `text` as `filename`. */
export function downloadTextFile(
  text: string,
  filename: string,
  mime = "text/csv;charset=utf-8;",
): void {
  const blob = new Blob([text], { type: mime });
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = filename;
  link.click();
  URL.revokeObjectURL(url);
}

File: LDS ADMIN/app/csvexport.py
"""Aligned, RFC 4180 CSV writing.

A deliberate port of the browser helper in Stratus (client/src/lib/csvExport.ts)
so a client who exports weather data from Stratus and calibration data from this
panel gets files that look and parse the same way. The two products previously
disagreed: one padded its columns, the other joined raw values with commas.

Two alignment rules, both chosen for a reason:

  - Padding goes on the RIGHT, so a value still begins immediately after its
    comma. Spaces inside a field are part of the field under RFC 4180, and a
    numeric column with LEADING spaces is exactly what makes Excel import that
    column as text. Trailing spaces are dropped when a cell is coerced to a
    number, so the file stays machine-readable as well as legible.
  - The last column is never padded. Trailing whitespace at end of line buys no
    alignment and several diff and lint tools flag it.

Width is computed per block, not across the document. A calibration record holds
several tables with different shapes, and stretching a two-column summary to the
width of a five-column log would misalign both.

A UTF-8 BOM is prepended once, at the top of the document. Without it Excel
assumes the system code page.

Property of METRON (PTY) LTD | Inteltronics
Developed by L.J. Esterhuizen, Inteltronics
"""

BOM = "\ufeff"

# Characters that force a field to be quoted, per RFC 4180.
_MUST_QUOTE = ('"', ",", "\n", "\r")


def csv_field(value):
    """Quote one value to RFC 4180. None becomes an empty field."""
    text = "" if value is None else str(value)
    if any(ch in text for ch in _MUST_QUOTE):
        return '"' + text.replace('"', '""') + '"'
    return text


def align_block(rows):
    """Format a grid of already-stringified rows into aligned CSV lines.

    `rows` is a sequence of sequences. Ragged rows are allowed: a short row is
    treated as having empty trailing fields, which is what a section title row
    is.
    """
    cells = [[csv_field(v) for v in row] for row in rows]
    if not cells:
        return []

    col_count = max(len(r) for r in cells)
    widths = [0] * col_count
    for row in cells:
        # A one-cell row is a section title, not a table row, so it must not set
        # the width of column A. Otherwise a heading like "Oscillator
        # recalibration by trigger" padded every label under it out to 35
        # characters and the table it introduces looked worse than unaligned.
        if len(row) < 2:
            continue
        for i, cell in enumerate(row):
            if len(cell) > widths[i]:
                widths[i] = len(cell)

    lines = []
    for row in cells:
        last = len(row) - 1
        out = []
        for i, cell in enumerate(row):
            # Never pad the final field of the row it is actually in, otherwise a
            # one-cell title row would carry the whole grid's width in spaces.
            out.append(cell if i == last else cell.ljust(widths[i]))
        lines.append(",".join(out))
    return lines


def build_aligned_csv(headers, rows, extra_rows=()):
    """One aligned table as a complete CSV document.

    `extra_rows` are appended after a blank line and are NOT padded to the grid:
    provenance lines are prose, not table rows.
    """
    lines = align_block([list(headers)] + [list(r) for r in rows])
    if extra_rows:
        lines.append("")
        lines.extend(str(r) for r in extra_rows)
    return BOM + "\n".join(lines) + "\n"


def build_csv_document(blocks, notes=()):
    """Several aligned tables as one CSV document.

    `blocks` is a sequence of row-sequences; each block is aligned on its own and
    separated from the next by a blank line. Empty blocks are skipped so a
    section that had no data does not leave a gap.

    `notes` are plain single-field lines appended at the end, each wrapped in its
    own row so a spreadsheet keeps the text in column A.
    """
    out = []
    for rows in blocks:
        lines = align_block([list(r) for r in rows])
        if not lines:
            continue
        if out:
            out.append("")
        out.extend(lines)
    if notes:
        if out:
            out.append("")
        out.extend(csv_field(n) for n in notes)
    return BOM + "\n".join(out) + "\n"
